SU(n) group

juli31 opened the pull request.

juli31 committed:
Create SU.py

juli31 committed:
Update SU.py

jksuom commented on sympy/liealgebras/SU.py:
```diff
@@ -0,0 +1,74 @@
+from sympy.matrices import zeros, ones, Matrix, randMatrix, matrices
+from sympy import mpmath as mp
```
mpmath is no longer a part of SymPy. This should probably be `import mpmath`.

jksuom commented on sympy/liealgebras/SU.py:
```diff
@@ -0,0 +1,74 @@
+from sympy.matrices import zeros, ones, Matrix, randMatrix, matrices
+from sympy import mpmath as mp
+from mpmath import qr_solve
```
Are qr_solve and lu_solve from mpmath essential? There should be also SymPy versions available.

jksuom commented on sympy/liealgebras/SU.py:
```diff
@@ -0,0 +1,74 @@
+from mpmath import qr_solve
+from mpmath import lu_solve
+
+#generator of SU(n) group
```
This (or part of it) could belong the docstring. Similarly with other comments below.

jksuom commented on sympy/liealgebras/SU.py:
```diff
@@ -0,0 +1,74 @@
+#here
+
+def SU_generator(n,str):
+    generator=[]
```
Please add a space on either side of =.

jksuom commented on sympy/liealgebras/SU.py:
```diff
@@ -0,0 +1,74 @@
+def SU_generator(n,str):
+    generator=[]
+    list_of_generators=[]
+    if (n==2):
```
A space on either side of ==.

jksuom commented on sympy/liealgebras/SU.py:
```diff
@@ -0,0 +1,74 @@
+    generator=[]
+    list_of_generators=[]
+    if (n==2):
+        generator.append(Matrix([[0,0+1j],[0+1j,0]]))
```
A space after each comma.

jksuom commented on sympy/liealgebras/SU.py:
```diff
@@ -0,0 +1,74 @@
+    list_of_generators=[]
+    if (n==2):
+        generator.append(Matrix([[0,0+1j],[0+1j,0]]))
+        generator.append(Matrix([[0,0-1j],[0+1j,0]]))
```
Why `0 - 1`, `0 + 1` instead of -1 and 1?

jksuom commented:
Trailing whitespace can be found by running `bin/doctest SU.py`.

asmeurer commented on sympy/liealgebras/SU.py:
```diff
@@ -0,0 +1,74 @@
+#used for every SU(n) group, i will have to write only a new generator
+#here
+
+def SU_generator(n,str):
```
Call the argument something other than `str`. `str` is a builtin name in Python and it's bad to shadow those.

jksuom commented:
It seems that this is concerned with the Lie algebra of SU(n), not the group itself. It is customary to use lower case letters to denote Lie algebras, hence a more suitable file name would be su.py. The name can be changed by `git mv`.

juli31 committed:
Update and rename SU.py to su.py

Necessary code style changes

juli31 commented:
Thank you for your comments. Can you see if all the necessary changes were made?

asmeurer commented:
Capital letters in module names also cause issues on systems with case insensitive file systems like macOS and Windows, and should be avoided anyway.

juli31 committed:
Update su.py

juli31 committed:
Update su.py

Generator for all su(n) algebras was added, su function was removed and I added random_su_element function instead of it.

juli31 committed:
Create SU.py

juli31 commented:
I think I fixed all bugs but Travis CI still reports failure of some tests which are not related to my code. Please tell me if I'm wrong.

jksuom commented:
It looks like your code tries to invert a random matrix that is not invertible while running LUsolve.

BTW, there are methods in SymPy that can be used to determine if a matrix is unitary (like the elements of the group SU(n)) or antihermitian (like the elements of the Lie algebra su(n)). So there is no need for new functions for those purposes.

czgdp1807 commented:
Closed accidentally. @jksuom Is this PR still needed?

jksuom commented:
It seems that `su_generator` will create a diagonal matrices of trace 0 and non-diagonal elementary matrices with two non-zero entries. The construction of such matrices by two loops of row_joins and col_joins looks inefficient. It would probably suffice to start with a matrix of zeros and then set the two entries at (i, j) and (j, i).

The doctring of `is_su` speaks of a matrix of the *group* but in the test a Lie algebra element is entered. The function usually fails with ValueError as most of the Lie algebra matrices are not invertible. However, the function is not needed as `.is_hermitian()` can be used instead.

The `liealgebra` module does not currently handle the actual algebra structures, but I don't think this PR would add much to it.

czgdp1807 commented:
Closing as the PR is inactive from quite a long time and the above comment says(according to my interpretation) doesn't add much to `master`. However if you think this can be re-opened then please let us know. Thanks for your contributions.
